libkatana_python_native/src/PropertyGraph.cpp

```diff
@@ -0,0 +1,742 @@
+      py::arg("table") = py::none(),
+      py::arg("txn_ctx") = TxnContextArgumentHandler::default_value);
+  cls.def(
+      "upsert_edge_property",
+      [](PropertyGraph& self, const py::object& table, TxnContext* txn_ctx,
+         const py::kwargs& kwargs) -> Result<void> {
+        std::shared_ptr<arrow::Table> arrow_table =
+            KATANA_CHECKED(PythonArgumentsToTable(table, kwargs));
+        py::gil_scoped_release guard;
+        TxnContextArgumentHandler txn_handler(txn_ctx);
+        return self.UpsertEdgeProperties(arrow_table, txn_handler.get());
+      },
+      py::arg("table") = py::none(),
+      py::arg("txn_ctx") = TxnContextArgumentHandler::default_value);
+
+  cls.def(
+      "remove_edge_property",
+      [](PropertyGraph& self, const std::string& name, TxnContext* txn_ctx) {
+        TxnContextArgumentHandler txn_handler(txn_ctx);
+        return self.RemoveEdgeProperty(name, txn_handler.get());
+      },
+      py::arg("name"),
+      py::arg("txn_ctx") = TxnContextArgumentHandler::default_value,
+      py::call_guard<py::gil_scoped_release>());
+
+  cls.def("has_node_index", &PropertyGraph::HasNodeIndex, py::arg("name"));
+  cls.def(
+      "get_node_index",
+      [](PropertyGraph& self, const std::string& name)
+          -> Result<std::shared_ptr<
+              katana::EntityIndex<katana::GraphTopology::Node>>> {
+        if (!self.HasNodeIndex(name)) {
+          KATANA_CHECKED(self.MakeNodeIndex(name));
+        }
+        return self.GetNodeIndex(name);
+      },
+      py::arg("name"), py::return_value_policy::reference_internal);
+  cls.def("has_edge_index", &PropertyGraph::HasEdgeIndex, py::arg("name"));
+  cls.def(
+      "get_edge_index",
+      [](PropertyGraph& self, const std::string& name)
+          -> Result<std::shared_ptr<
+              katana::EntityIndex<katana::GraphTopology::Edge>>> {
+        if (!self.HasEdgeIndex(name)) {
+          KATANA_CHECKED(self.MakeEdgeIndex(name));
+        }
+        return self.GetEdgeIndex(name);
+      },
+      py::arg("name"), py::return_value_policy::reference_internal);
+
+  cls.def("unload_topologies", &PropertyGraph::DropAllTopologies);
+
+  cls.def(
+      "write",
+      [](PropertyGraph& self, const std::string* path,
+         const std::string& provenance) {
+        if (path) {
+          return self.Write(*path, provenance);
+        }
+        return self.Commit(provenance);
+      },
+      py::arg("path") = py::none(),
+      py::arg("provenance") = py::str("katana.local"),
+      py::call_guard<py::gil_scoped_release>(),
+      R"""(
+      Write the property graph to the specified path or URL (or the original path it was loaded from if path is
+      not provided). Provide lineage information in the form of a command line.
+
+      :param path: The path to which to write or None to use `self.path`.
+      :type path: str or Path
+      :param command_line: Lineage information in the form of a command line.
+      :type command_line: str
```

Comment:
Will this string be formatted correctly?

Reply:
I'm not aware of any formatting requirements. I honestly think we need to get ride of it. I think it has been subsumed by the op log.